File: tools/vector-gen/src/main/java/cash/freer/mac/vectorgen/BchSchnorr.java
package cash.freer.mac.vectorgen;

import org.bitcoinj.core.Utils;

import java.math.BigInteger;
import java.security.MessageDigest;

/**
 * BCH-2019 Schnorr signature — the variant freecashj ships and FC-AJDK uses
 * (see FC-AJDK/.../core/fch/SchnorrSignature.java). Pre-dates BIP-340; not
 * interoperable with it.
 *
 * Sign:
 *   k0 = SHA-256(d || m) mod n
 *   R  = k0 · G
 *   k  = (jacobi(R.y) == 1) ? k0 : (n - k0)
 *   e  = SHA-256(R.x || P_compressed_33 || m) mod n
 *   s  = (e·d + k) mod n
 *   sig = R.x(32) || s(32)
 *
 * Verify:
 *   r,s from sig
 *   e  = SHA-256(r || P_compressed_33 || m) mod n
 *   R  = s·G + (n - e)·P
 *   valid iff jacobi(R.y) == 1 and r == R.x
 *
 * Naive textbook EC math via BigInteger — slow but faithful to the
 * reference. Used only for test-vector generation.
 */
public final class BchSchnorr {

    public static final BigInteger P = new BigInteger(
            "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F", 16);
    public static final BigInteger N = new BigInteger(
            "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", 16);
    public static final BigInteger[] G = {
            new BigInteger("79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798", 16),
            new BigInteger("483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8", 16)
    };
    private static final BigInteger TWO = BigInteger.valueOf(2);
    private static final BigInteger THREE = BigInteger.valueOf(3);

    public static byte[] sign(byte[] msg, BigInteger seckey) {
        if (msg.length != 32) {
            throw new IllegalArgumentException("msg must be 32 bytes");
        }
        if (seckey.signum() <= 0 || seckey.compareTo(N.subtract(BigInteger.ONE)) > 0) {
            throw new IllegalArgumentException("seckey out of range");
        }

        try {
            byte[] nonceInput = new byte[64];
            System.arraycopy(to32(seckey), 0, nonceInput, 0, 32);
            System.arraycopy(msg, 0, nonceInput, 32, 32);
            BigInteger k0 = toBigInt(sha256(nonceInput)).mod(N);
            if (k0.signum() == 0) {
                throw new IllegalStateException("k0 == 0 (negligible probability)");
            }

            BigInteger[] rPoint = pointMul(G, k0);
            BigInteger k = jacobi(rPoint[1]).equals(BigInteger.ONE) ? k0 : N.subtract(k0);

            byte[] rxBytes = to32(rPoint[0]);
            byte[] challengeInput = new byte[32 + 33 + 32];
            System.arraycopy(rxBytes, 0, challengeInput, 0, 32);
            System.arraycopy(compressPoint(pointMul(G, seckey)), 0, challengeInput, 32, 33);
            System.arraycopy(msg, 0, challengeInput, 65, 32);
            BigInteger e = toBigInt(sha256(challengeInput)).mod(N);

            byte[] out = new byte[64];
            System.arraycopy(rxBytes, 0, out, 0, 32);
            System.arraycopy(to32(e.multiply(seckey).add(k).mod(N)), 0, out, 32, 32);
            return out;
        } catch (Exception e) {
            throw new RuntimeException("schnorr_sign failed", e);
        }
    }

    public static boolean verify(byte[] msg, byte[] pubkey, byte[] sig) {
        if (msg.length != 32 || pubkey.length != 33 || sig.length != 64) {
            throw new IllegalArgumentException("bad lengths");
        }
        BigInteger[] point = decompressPoint(pubkey);
        if (point == null) return false;
        BigInteger r = toBigInt(sig, 0, 32);
        BigInteger s = toBigInt(sig, 32, 32);
        if (r.compareTo(P) >= 0 || s.compareTo(N) >= 0) return false;

        try {
            byte[] challengeInput = new byte[32 + 33 + 32];
            System.arraycopy(sig, 0, challengeInput, 0, 32);
            System.arraycopy(compressPoint(point), 0, challengeInput, 32, 33);
            System.arraycopy(msg, 0, challengeInput, 65, 32);
            BigInteger e = toBigInt(sha256(challengeInput)).mod(N);

            BigInteger[] rPoint = pointAdd(pointMul(G, s), pointMul(point, N.subtract(e)));
            return rPoint != null
                    && jacobi(rPoint[1]).equals(BigInteger.ONE)
                    && r.equals(rPoint[0]);
        } catch (Exception e) {
            throw new RuntimeException("schnorr_verify failed", e);
        }
    }

    // --- EC math --------------------------------------------------------

    public static BigInteger[] pointAdd(BigInteger[] a, BigInteger[] b) {
        if (a == null) return b;
        if (b == null) return a;
        if (a[0].equals(b[0]) && !a[1].equals(b[1])) return null;

        BigInteger lam;
        if (a[0].equals(b[0]) && a[1].equals(b[1])) {
            lam = THREE.multiply(a[0]).multiply(a[0])
                    .multiply(TWO.multiply(a[1]).modPow(P.subtract(TWO), P)).mod(P);
        } else {
            lam = b[1].subtract(a[1])
                    .multiply(b[0].subtract(a[0]).modPow(P.subtract(TWO), P)).mod(P);
        }
        BigInteger x3 = lam.multiply(lam).subtract(a[0]).subtract(b[0]).mod(P);
        BigInteger y3 = lam.multiply(a[0].subtract(x3)).subtract(a[1]).mod(P);
        return new BigInteger[]{x3, y3};
    }

    public static BigInteger[] pointMul(BigInteger[] p, BigInteger k) {
        BigInteger[] r = null;
        BigInteger[] cur = p;
        for (int i = 0; i < 256; i++) {
            if (k.shiftRight(i).and(BigInteger.ONE).equals(BigInteger.ONE)) {
                r = pointAdd(r, cur);
            }
            cur = pointAdd(cur, cur);
        }
        return r;
    }

    public static BigInteger jacobi(BigInteger x) {
        return x.modPow(P.subtract(BigInteger.ONE).divide(TWO), P);
    }

    public static byte[] compressPoint(BigInteger[] point) {
        byte[] out = new byte[33];
        out[0] = point[1].testBit(0) ? (byte) 0x03 : (byte) 0x02;
        System.arraycopy(to32(point[0]), 0, out, 1, 32);
        return out;
    }

    public static BigInteger[] decompressPoint(byte[] b) {
        if (b[0] != 0x02 && b[0] != 0x03) return null;
        boolean odd = (b[0] == 0x03);
        BigInteger x = toBigInt(b, 1, 32);
        BigInteger ySq = x.modPow(THREE, P).add(BigInteger.valueOf(7)).mod(P);
        BigInteger y0 = ySq.modPow(P.add(BigInteger.ONE).divide(BigInteger.valueOf(4)), P);
        if (!ySq.equals(y0.modPow(TWO, P))) return null;
        BigInteger y = (y0.testBit(0) != odd) ? P.subtract(y0) : y0;
        return new BigInteger[]{x, y};
    }

    // --- utilities ------------------------------------------------------

    private static byte[] to32(BigInteger value) {
        return Utils.bigIntegerToBytes(value, 32);
    }

    private static BigInteger toBigInt(byte[] data) {
        return new BigInteger(1, data);
    }

    private static BigInteger toBigInt(byte[] data, int offset, int len) {
        byte[] slice = new byte[len];
        System.arraycopy(data, offset, slice, 0, len);
        return new BigInteger(1, slice);
    }

    private static byte[] sha256(byte[] input) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(input);
        } catch (java.security.NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    private BchSchnorr() {}
}

File: Packages/FCCore/Sources/FCCore/Crypto/BchSchnorr.swift
import Foundation
import BigInt
import P256K

/// BitcoinCash 2019 Schnorr signature — the pre-BIP-340 variant used by
/// freecashj. Algorithm ported from
/// `FC-AJDK/.../core/fch/SchnorrSignature.java`, which itself is a port
/// of https://github.com/miketwk/bip-schnorr-java.
///
/// **Not interoperable with BIP-340.** Key differences from BIP-340:
/// - Nonce: `SHA-256(d || m)` (no tagged hash, no auxiliary data).
/// - Challenge: `SHA-256(R.x || P_compressed_33 || m)` (no tags, 33-byte
///   compressed pubkey, not 32-byte x-only).
/// - R selection: Y is a quadratic residue (`jacobi(R.y) == 1`), not
///   "even Y".
///
/// **Implementation notes.** EC point multiplications run through
/// libsecp256k1 (`secp256k1_ec_pubkey_create`, `_tweak_mul`,
/// `_seckey_tweak_*`), so signing and verification each take a few
/// milliseconds rather than the ~30 s the previous BigInt-only path
/// took. The remaining BigInt work is a single mod-n reduction per
/// SHA-256 output and the jacobi mod-p exponentiation on `R.y` —
/// roughly 256 BigInt squarings, each microseconds.
///
/// The wire format is byte-identical to the prior BigInt version (the
/// `BchSchnorrTests` byte-exact parity vectors against freecashj still
/// pass).
public enum BchSchnorr {

    public enum Failure: Error, CustomStringConvertible {
        case invalidMessageLength
        case invalidSeckey
        case invalidPubkey
        case invalidSignatureLength
        case signingFailed

        public var description: String {
            switch self {
            case .invalidMessageLength:   return "BchSchnorr: message must be 32 bytes"
            case .invalidSeckey:          return "BchSchnorr: seckey out of range"
            case .invalidPubkey:          return "BchSchnorr: pubkey must be 33-byte compressed"
            case .invalidSignatureLength: return "BchSchnorr: signature must be 64 bytes"
            case .signingFailed:          return "BchSchnorr: signing failed"
            }
        }
    }

    // MARK: - secp256k1 constants

    /// Field prime — used for the jacobi(R.y) check (mod-p exponentiation).
    static let p = BigInt("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F", radix: 16)!
    /// Curve order — used to reduce SHA-256 outputs into valid scalars.
    static let n = BigInt("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", radix: 16)!

    // MARK: - Public API

    /// Sign a 32-byte `message` (typically a sighash) with `privateKey`.
    /// Returns a 64-byte signature in the BCH 2019 Schnorr wire layout
    /// `R.x || s`. Deterministic: same `(privateKey, message)` always
    /// yields the same bytes.
    public static func sign(message: Data, privateKey: Data) throws -> Data {
        guard message.count == 32 else { throw Failure.invalidMessageLength }
        guard privateKey.count == 32 else { throw Failure.invalidSeckey }

        // Validate privkey via libsecp256k1 (1 <= d < n).
        let dKey: P256K.Signing.PrivateKey
        do {
            dKey = try P256K.Signing.PrivateKey(dataRepresentation: privateKey)
        } catch {
            throw Failure.invalidSeckey
        }

        // 1) k0 = SHA-256(d || m) reduced mod n. SHA-256 output is
        //    almost always already < n; reduction handles the
        //    ~2⁻¹²⁸ tail.
        var nonceInput = Data(capacity: 64)
        nonceInput.append(privateKey)
        nonceInput.append(message)
        let k0Bytes = try reduceModN(Hash.sha256(nonceInput))

        // 2) R = k0 · G. Treat k0 as a privkey; libsecp256k1 derives
        //    the corresponding pubkey, which is exactly k0·G.
        let k0Key: P256K.Signing.PrivateKey
        do {
            k0Key = try P256K.Signing.PrivateKey(dataRepresentation: k0Bytes)
        } catch {
            throw Failure.signingFailed
        }
        let rUncompressed = k0Key.publicKey.uncompressedRepresentation
        guard rUncompressed.count == 65 else { throw Failure.signingFailed }
        let rx = rUncompressed.subdata(in: 1..<33)
        let ry = rUncompressed.subdata(in: 33..<65)

        // 3) BCH Schnorr canonicalization: pick the (x, y) where
        //    jacobi(y) == 1. If the natural y from k0·G fails the
        //    jacobi test, flip k0 → n - k0 (which gives the same x
        //    with the opposite y).
        let kKey = (jacobiIsOne(bigInt(ry))) ? k0Key : k0Key.negation
        let kBytes = kKey.dataRepresentation

        // 4) e = SHA-256(R.x || P_compressed(33) || m) reduced mod n.
        let pCompressed = dKey.publicKey.dataRepresentation
        guard pCompressed.count == 33 else { throw Failure.signingFailed }
        var challenge = Data(capacity: 32 + 33 + 32)
        challenge.append(rx)
        challenge.append(pCompressed)
        challenge.append(message)
        let eBytes = try reduceModN(Hash.sha256(challenge))

        // 5) s = (e * d + k) mod n via the seckey-tweak primitives.
        //    Both operations fail iff the result is zero, which has
        //    negligible probability for random k0/e — surface as
        //    `signingFailed` if it ever happens.
        let edKey: P256K.Signing.PrivateKey
        let sKey: P256K.Signing.PrivateKey
        do {
            edKey = try dKey.multiply(Array(eBytes))
            sKey = try edKey.add(Array(kBytes))
        } catch {
            throw Failure.signingFailed
        }

        var out = Data(capacity: 64)
        out.append(rx)
        out.append(sKey.dataRepresentation)
        return out
    }

    /// Verify a 64-byte BCH Schnorr signature against a 33-byte
    /// compressed pubkey and a 32-byte message. Returns false (rather
    /// than throwing) for any cryptographic mismatch; only the input
    /// shape errors throw.
    public static func verify(message: Data, publicKey: Data, signature: Data) throws -> Bool {
        guard message.count == 32 else { throw Failure.invalidMessageLength }
        guard publicKey.count == 33 else { throw Failure.invalidPubkey }
        guard signature.count == 64 else { throw Failure.invalidSignatureLength }

        // Parse pubkey. An invalid encoding (wrong prefix byte, off-curve x)
        // is a verification failure, not a programmer error.
        let pubKey: P256K.Signing.PublicKey
        do {
            pubKey = try P256K.Signing.PublicKey(dataRepresentation: publicKey, format: .compressed)
        } catch {
            return false
        }

        let sig = Data(signature)
        let rBytes = sig.prefix(32)
        let sBytes = sig.suffix(32)
        let r = bigInt(rBytes)
        let s = bigInt(sBytes)
        // r < p (field), s < n (curve order). r == 0 isn't excluded
        // by the spec — leave to the final x-equality check to reject.
        guard r < p, s < n, s.signum() > 0 else { return false }

        // 1) e = SHA-256(R.x || P_compressed || m) mod n.
        var challenge = Data(capacity: 32 + 33 + 32)
        challenge.append(rBytes)
        challenge.append(publicKey)
        challenge.append(message)
        let eBytes: Data
        do {
            eBytes = try reduceModN(Hash.sha256(challenge))
        } catch {
            return false
        }

        // 2) Reconstruct R = s·G + (n - e)·P  ≡  s·G - e·P.
        do {
            let sKey = try P256K.Signing.PrivateKey(dataRepresentation: sBytes)
            let sG = sKey.publicKey
            let ePNeg = try pubKey.multiply(Array(eBytes), format: .compressed).negation
            let reconstructed = try sG.combine([ePNeg], format: .uncompressed)
            let raw = reconstructed.uncompressedRepresentation
            guard raw.count == 65 else { return false }
            let xBytes = raw.subdata(in: 1..<33)
            let yBytes = raw.subdata(in: 33..<65)

            // 3) Accept iff jacobi(R.y) == 1 AND R.x == r.
            return jacobiIsOne(bigInt(yBytes)) && bigInt(xBytes) == r
        } catch {
            return false
        }
    }

    // MARK: - helpers

    /// Reduce a 32-byte big-endian integer to the canonical
    /// `[1, n-1]` range. Returns the 32-byte big-endian representation
    /// of `value mod n`. Throws ``Failure/signingFailed`` if the
    /// reduced value is zero (negligible probability for SHA-256
    /// outputs, but a hard signing failure when it occurs).
    static func reduceModN(_ raw: Data) throws -> Data {
        let reduced = bigInt(raw).modulus(n)
        guard reduced.signum() > 0 else { throw Failure.signingFailed }
        return to32(reduced)
    }

    /// True iff `y` is a quadratic residue mod `p`. For secp256k1
    /// (where `p ≡ 3 (mod 4)`) this is `y^((p-1)/2) mod p == 1`.
    static func jacobiIsOne(_ y: BigInt) -> Bool {
        y.power((p - 1) / 2, modulus: p) == BigInt(1)
    }

    static func bigInt(_ data: Data) -> BigInt {
        BigInt(sign: .plus, magnitude: BigUInt(Data(data)))
    }

    static func to32(_ value: BigInt) -> Data {
        let raw = value.magnitude.serialize()
        if raw.count == 32 { return raw }
        if raw.count > 32 { return raw.suffix(32) }
        var padded = Data(repeating: 0, count: 32 - raw.count)
        padded.append(raw)
        return padded
    }
}
